File: destack-py/destack/encoder/cson/encoder.py
import json
from typing import Any, ClassVar, override

from destack.language.core import (
    BinaryReader,
    BinaryWriter,
    BuiltinObject,
    Cson,
    Encoder,
    Encoding,
    NodeType,
    ObjectKind,
    Session,
    StructType,
    Type,
)

from .generate import CSON_OBJECT_ENCODERS
from .wiring import pack_cson, unpack_cson


class CsonEncoder(Encoder[Cson]):
    """Encoder for our custom constant folded JSON format."""

    encoding: ClassVar[Encoding] = Encoding.CSON

    @override
    def pack_object(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        object: BuiltinObject,
    ) -> Cson:
        encoder = CSON_OBJECT_ENCODERS.get((kind, metatype))
        assert encoder is not None, f"no CsonObjectEncoder for {kind.name}:{metatype.name}"
        return encoder.pack_object(object)

    @override
    def pack_object_binary(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        object: BuiltinObject,
        writer: BinaryWriter,
    ) -> None:
        encoder = CSON_OBJECT_ENCODERS.get((kind, metatype))
        assert encoder is not None, f"no CsonObjectEncoder for {kind.name}:{metatype.name}"
        object_packed = encoder.pack_object(object)
        writer.write_bytes(json.dumps(object_packed).encode("utf-8"))

    @override
    def unpack_object(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        value: Cson,
        session: Session | None,
    ) -> BuiltinObject:
        encoder = CSON_OBJECT_ENCODERS.get((kind, metatype))
        assert encoder is not None, f"no CsonObjectEncoder for {kind.name}:{metatype.name}"
        return encoder.unpack_object(value, session)

    @override
    def unpack_object_binary(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        reader: BinaryReader,
        session: Session | None,
    ) -> BuiltinObject:
        encoder = CSON_OBJECT_ENCODERS.get((kind, metatype))
        assert encoder is not None, f"no CsonObjectEncoder for {kind.name}:{metatype.name}"
        value_decoded = json.loads(reader.read_bytes().decode("utf-8"))
        return encoder.unpack_object(value_decoded, session)

    @override
    def pack_value(
        self,
        type: Type,
        value: Any,
    ) -> Cson:
        return pack_cson(value, type)

    @override
    def pack_value_binary(
        self,
        type: Type,
        value: Any,
        writer: BinaryWriter,
    ) -> None:
        writer.write_bytes(pack_cson(value, type).encode("utf-8"))

    @override
    def unpack_value(
        self,
        type: Type,
        value: Cson,
        session: Session | None,
    ) -> Any:
        return unpack_cson(value, type, session)

    @override
    def unpack_value_binary(
        self,
        type: Type,
        reader: BinaryReader,
        session: Session | None,
    ) -> Any:
        value_decoded = json.loads(reader.read_bytes().decode("utf-8"))
        return unpack_cson(value_decoded, type, session)

File: destack-ts/src/encoder/cson/encoder.ts
import { CSON_OBJECT_ENCODERS, getObjectKey } from "@destack/encoder/cson/generate";
import { loadEncoders } from "@destack/encoder/cson/generated";
import { packCson, unpackCson } from "@destack/encoder/cson/wiring";
import {
  type BinaryReader,
  type BinaryWriter,
  type BuiltinObject,
  type Encoder,
  NodeType,
  ObjectKind,
  type Session,
  StructType,
  type Type,
} from "@destack/language";

/** Encoder for our CSON format. */
export class CsonEncoder implements Encoder<any> {
  constructor() {
    loadEncoders();
  }

  packObject(kind: ObjectKind, metatype: NodeType | StructType, object: BuiltinObject): any {
    const encoder = CSON_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no CsonEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.packObject(object);
  }

  packObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    object: BuiltinObject,
    writer: BinaryWriter,
  ): void {
    const encoder = CSON_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no CsonEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    const objectPacked = encoder.packObject(object);
    writer.writeJson(objectPacked);
  }

  unpackObject(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    value: any,
    session: Session | null,
  ): BuiltinObject {
    const encoder = CSON_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no CsonEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.unpackObject(value, session);
  }

  unpackObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    reader: BinaryReader,
    session: Session | null,
  ): BuiltinObject {
    const encoder = CSON_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no CsonEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    const objectPacked = reader.readJson();
    return encoder.unpackObject(objectPacked, session);
  }

  packValue(value: any, type: Type): any {
    const cson = packCson(value, type);
    return cson;
  }

  packValueBytes(value: any, type: Type, writer: BinaryWriter): void {
    const cson = packCson(value, type);
    writer.writeJson(cson);
  }

  unpackValue(type: Type, value: any, session: Session | null): any {
    const cson = unpackCson(value, type);
    return cson;
  }

  unpackValueBytes(type: Type, reader: BinaryReader, session: Session | null): any {
    const cson = unpackCson(reader.readJson(), type);
    return cson;
  }
}
